### PR TITLE
feat: extend function `fraction` with support for units (see #2918)

### PR DESCRIPTION
**Pull request by josdejong**
Implements `math.fraction(unit)`, which converts the numeric value of the `Unit` to `Fraction`.

See: #2918, #2921

@brunoSnoww can you have a look at this PR?

### CODE REVIEW

File: src/type/fraction/function/fraction.js [hunk 1]
```diff
@@ -37,7 +37,7 @@ export const createFraction = /* #__PURE__ */ factory(name, dependencies, ({ typ
    *
    *    bignumber, number, string, unit
    *
-   * @param {number | string | Fraction | BigNumber | Array | Matrix} [args]
+   * @param {number | string | Fraction | BigNumber | Unit | Array | Matrix} [args]
```
Comment (brunoSnoww) [inline src/type/fraction/function/fraction.js:40]: the ```fraction``` type in the index.d.ts file needs to be update as well :))

**Review by brunoSnoww — APPROVED** [review]
LGTM, just some little comments. Also a PR to update the docs about Units to explain that the built-in units always have a number as value would be nice 🙃

File: src/type/fraction/function/fraction.js (hunk 1 above)
Comment (josdejong) [reply src/type/fraction/function/fraction.js:40]: Good point, I've updated the TypeScript definitions

File: src/type/fraction/function/fraction.js [hunk 2]
```diff
@@ -71,6 +71,12 @@ export const createFraction = /* #__PURE__ */ factory(name, dependencies, ({ typ
       return x // fractions are immutable
     },
 
+    Unit: typed.referToSelf(self => (x) => {
+      const clone = x.clone()
+      clone.value = self(x.value)
+      return clone
+    }),
```
[the remark this answers is not in the document]
Comment (josdejong) [reply src/type/fraction/function/fraction.js:78]: `new Fraction(...)` will work when the value is a `number` but not when it is a `BigNumber`, therefore I reference the `fraction(...)` function itself.

File: docs/datatypes/units.md [hunk 3]
```diff
@@ -203,6 +203,26 @@ math.createUnit('θ', '1 rad')
...
+```
+
+When using the expression parser, it is possible to configure numeric values to be parsed as `Fraction` or `BigNumber`:
+
+```js
+math.config({ number: 'Fraction' })
+math.evaluate('10 inch').toNumeric('cm') // Fraction 127/5
+```
+
+
```
Comment (brunoSnoww) [inline docs/datatypes/units.md:225]: What you think about adding the following use case ? 

_Beware that converting irrational numbers to it's fraction representation will lead to a loss of precision in comparison to it's numerical value_
```javascript 
const a = math.unit(math.pi);
const fa = math.fraction(a) 
const isEqual = fa.equals(a)) // it throws: loss of precision
```

**Comment by brunoSnoww** [comment]
Good point. For consistency we should indeed give such support for ```number``` and ```big number```. For the case of ```number```, what do you think of the following, since we want a type of number as an output ? 

```javascript
Unit: function (unit) {
  return unit.toNumber(unit.formatUnits())
},
```

so that both are the same:

```javascript 
const a1 = math.number(math.unit(45,'cm'))
const a2 = math.number(math.unit(45,'cm'),'cm')
a1 === a2 // true
```

File: docs/datatypes/units.md (hunk 3 above)
Comment (josdejong) [reply docs/datatypes/units.md:225]: I think that throwing an error in this case is fine.

The idea behind it is: when some operation like `equals` _implicitly_ (internally, silently) converts a fraction and that results in a loss of precision, it should throw an error. When the user is OK with that, he/she can _explicitly_ convert the number into a fraction using `math.fraction` and that should not throw an error. Does that make sense?

**Comment by josdejong** [comment]
> Good point. For consistency we should indeed give such support for `number` and `big number`. For the case of `number`, what do you think of the following, since we want a type of number as an output ?
> 
> ```js
> Unit: typed.referToSelf(self => x => {
>     return self(x, x.formatUnits())
> }),
> ```
> 
> so that both are the same:
> 
> ```js
> const a1 = math.number(math.unit(45,'cm'))
> const a2 = math.number(math.unit(45,'cm'),'cm')
> a1 === a2 // 45 === 45 , true
> ```

I understand what you mean. I think it is best not to let `math.number(unit)` return a number, since it is not explicit what unit is used to convert it to a number. That is why it currently throws an error. If I add two units, one in `cm` and one in `inch`, and then call `math.number` on the result, will the output be a number in cm, or inch, or meter?

**Comment by brunoSnoww** [comment]
> > Good point. For consistency we should indeed give such support for `number` and `big number`. For the case of `number`, what do you think of the following, since we want a type of number as an output ?
> > ```js
> > Unit: typed.referToSelf(self => x => {
> >     return self(x, x.formatUnits())
> > }),
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > so that both are the same:
> > ```js
> > const a1 = math.number(math.unit(45,'cm'))
> > const a2 = math.number(math.unit(45,'cm'),'cm')
> > a1 === a2 // 45 === 45 , true
> > ```
> 
> I understand what you mean. I think it is best not to let `math.number(unit)` return a number, since it is not explicit what unit is used to convert it to a number. That is why it currently throws an error. If I add two units, one in `cm` and one in `inch`, and then call `math.number` on the result, will the output be a number in cm, or inch, or meter?

Agreed 😬 makes sense to keep the throw for this case

File: docs/datatypes/units.md (hunk 3 above)
Comment (brunoSnoww) [reply docs/datatypes/units.md:225]: Yes , agreed :)

Comment (josdejong) [reply docs/datatypes/units.md:225]: Thanks for your feedback! I'll update the branch and merge it.

**Comment by josdejong** [comment]
Published now in `v11.8.0`. Thanks for thinking along Bruno.

**Pull request merged**